src/lib.rs

```diff
@@ -54,53 +54,64 @@ pub struct Cow<'a, T: Beef + ?Sized + 'a> {
-
-    /// Convert `&mut T::Owned` to `*mut T`, stripping `capacity`.
-    unsafe fn owned_ptr(owned: &mut Self::Owned) -> NonNull<Self>;
+    /// Convert `T::Owned` to `NonNull<T>` and capacity.
+    /// Return `None` for `0` capacity.
+    fn owned_into_parts(owned: Self::Owned) -> (NonNull<Self>, Option<NonZeroUsize>);
 
     /// Rebuild `T::Owned` from `NonNull<T>` and `capacity`. This can be done by the likes
     /// of [`Vec::from_raw_parts`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts).
-    unsafe fn rebuild(ptr: NonNull<Self>, capacity: usize) -> Self::Owned;
+    unsafe fn owned_from_parts(ptr: NonNull<Self>, capacity: NonZeroUsize) -> Self::Owned;
 }
 
 unsafe impl Beef for str {
     #[inline]
-    fn capacity(owned: &String) -> Option<NonZeroUsize> {
-        NonZeroUsize::new(owned.capacity())
-    }
+    fn owned_into_parts(owned: String) -> (NonNull<str>, Option<NonZeroUsize>) {
+        // Convert to `String::into_raw_parts` once stabilized
+        let mut owned = mem::ManuallyDrop::new(owned);
+        let ptr = slice_from_raw_parts_mut(
+            owned.as_mut_ptr(),
+            owned.len(),
+        );
 
-    #[inline]
-    unsafe fn owned_ptr(owned: &mut String) -> NonNull<str> {
-        NonNull::new_unchecked(owned.as_mut_str() as *mut str)
+        (
+            unsafe { NonNull::new_unchecked(ptr as *mut str) },
+            NonZeroUsize::new(owned.capacity()),
+        )
     }
 
     #[inline]
-    unsafe fn rebuild(mut ptr: NonNull<Self>, capacity: usize) -> String {
+    unsafe fn owned_from_parts(mut ptr: NonNull<Self>, capacity: NonZeroUsize) -> String {
         String::from_utf8_unchecked(Vec::from_raw_parts(
             ptr.as_mut().as_mut_ptr(),
             ptr.as_mut().len(),
-            capacity,
+            capacity.get(),
         ))
     }
 }
 
 unsafe impl<T: Clone> Beef for [T] {
     #[inline]
-    fn capacity(owned: &Vec<T>) -> Option<NonZeroUsize> {
-        NonZeroUsize::new(owned.capacity())
-    }
+    fn owned_into_parts(owned: Vec<T>) -> (NonNull<[T]>, Option<NonZeroUsize>) {
+        // Convert to `Vec::into_raw_parts` once stabilized
+        let mut owned = mem::ManuallyDrop::new(owned);
+        let ptr = slice_from_raw_parts_mut(
+            owned.as_mut_ptr(),
+            owned.len(),
+        );
 
-    #[inline]
-    unsafe fn owned_ptr(owned: &mut Vec<T>) -> NonNull<[T]> {
-        NonNull::new_unchecked(owned.as_mut_slice() as *mut [T])
+        (
+            unsafe { NonNull::new_unchecked(ptr) },
+            NonZeroUsize::new(owned.capacity()),
+        )
     }
 
     #[inline]
-    unsafe fn rebuild(mut ptr: NonNull<Self>, capacity: usize) -> Vec<T> {
-        Vec::from_raw_parts(ptr.as_mut().as_mut_ptr(), ptr.as_mut().len(), capacity)
+    unsafe fn owned_from_parts(mut ptr: NonNull<Self>, capacity: NonZeroUsize) -> Vec<T> {
+        Vec::from_raw_parts(
+            ptr.as_mut().as_mut_ptr(),
```

Comment:
This should probably be `ptr.cast::<T>().as_raw()` to avoid manifesting (and deriving from) a temporary borrow.